Committing a Rolled-Back Object (#10899)
mendix/docs: content/en/docs/refguide/modeling/application-logic/microflows-and-nanoflows/microflows/error-handling-in-microflows.md

* **Custom without rollback**
* **Continue**

In summary, the four error handling options do the following:

| Option | Behavior |
| --- | --- |

> | Rollback (default) |  All changes are reverted to the state before the microflow started, the microflow aborts, and a system error message is shown. |

Reviewer: I think a system error message is only shown if the error is not caught using error handling on a higher level and is allowed to bubble up.

Author: Updated in ff31fe7c06a52b948f2ee63f92f2661fba363e70

Reviewer: Some feedback after our discussion:
>  All changes are reverted to the state before the microflow started, the microflow aborts, 

'the microflow': is it the microflow in which the error handling is defined or the microflow that is called by the Call Microflow action on which error handling is defined.

> and a system error message is shown unless custom error handling is added in a microflow which called this one

I would phrase it as cause -> effect, something like: 'if the error is not handled, a system error message is shown'

Author: Updated in c920e2e1b012ff60ec7f55c03d4895bdf448594d

Reviewer: The entry for 'Continue' now mentions:
> No error is logged or shown to the end-user.

After the latest changes, no other entry references error logging or dialogs. So, shouldn't we remove it here?

> | Custom with rollback |  All changes are reverted to the state before the microflow started, an error flow is followed, and the microflow's subsequent behavior depends on whether the error handling flow ends with an error or end event. |

Reviewer: 'All changes ... state before the microflow started' might be a bit ambiguous here.

In the following setup (client request triggers MF1, MF1 triggers MF2, MF2 triggers MF3, which fails)
```
request -> MF1
MF1: Start -> ... database stuff ... -> call MF2 Rollback -> End
MF2: Start -> ... more database stuff ... -> call MF3 Custom with rollback
MF3: FAIL
```
all database operations in 'database stuff' and 'more database stuff' are reverted. If MF1 did 'call MF2 Custom without rollback' instead, only 'more database stuff' is reverted.

Reviewer 2: > If MF1 did 'call MF2 Custom without rollback' instead, only 'more database stuff' is reverted.

But aren't we talking about `Custom with rollback` here?

Reviewer: Yes, I just wanted to indicate how the call MF2 error handling configuration influences to which state the database is rolled back. This, to indicate that you do not know to what level the data is rolled back in the case of nested transactions. It could be to the database state at the start of the microflow calling the failing action (i.e. the start of MF2) or a database state at another level entirely.

Author: But doesn't it always end up in the state before this MF is called - the error propagation might then change it again - but for an instant when this particular MF errors the rollback will only go as far as the calling microflow and then work out what to do with the error there?

Reviewer: Not sure if I fully understand, probably because I am not 100% sure about what you mean by state. In the following:
```
request -> MF1
MF1: S -> A1 -> Call MF2 Rollback -> E
MF2: S -> A2 -> Call MF3 Custom with rollback -> E
MF3: S -> FAIL
```
when the `Call MF3` fails, execution continues in its error flow (execution state) while the database resets to the last savepoint, which is all the way to the start of `MF1`.

So we should probably be careful with the word 'state' as well :-)

Author: Updated to simplify - customers will need to look at the examples for the full picture.
c920e2e1b012ff60ec7f55c03d4895bdf448594d

> | Custom without rollback |  Changes made before the error are kept, an error flow is followed, and the microflow's subsequent behavior depends on whether the error handling flow ends with an error or end event. |

Reviewer 3: It may be more accurate to say that "Changes made before the activity causing the error are kept". If you commit a list of objects and the third one fails, the changes for the first two objects will be rolled back before the error flow starts.

Author: Updated in 89a5f3336953ea3f5b8ba388e6b806e496d294c4

> | Continue |  The microflow proceeds as if no error occurred, keeping all changes, and no error message is logged or displayed to the end-user. |

Reviewer 3: Similarly, it keeps all changes from before the activity causing the error.

Author: Updated in 89a5f3336953ea3f5b8ba388e6b806e496d294c4

| Option | Behavior |
| --- | --- |
| Rollback (default) | Reverts all changes, aborts the microflow, and returns the error to the calling microflow. |

> | Custom with rollback | Reverts all changes and follows the error flow. |

Reviewer: > Reverts all changes

Should we make it explicit that this about _database_ changes and that it reverts changes to the state of the last savepoint? Using 'all' might indicate all changes in the request?

Author: Good idea:
```suggestion
| Rollback (default) | Reverts all database changes, aborts the microflow, and returns the error to the calling microflow. |
| Custom with rollback | Reverts all changes and follows the error flow. |
```

It is important to know that when a microflow is started by an end-user (for example, by clicking a button), a savepoint is always created at the very beginning of the microflow. This is the case for **Microflow 1** in all the following examples. Knowing how savepoints are created and function is key to understanding how error handling works.

> In the following subsections, we introduce the functionality of each error handling option using simple examples. In the examples we have an activity which generates the error. In a real app, however, an error can be generated by any activity. Any changes which happen within the activity which fails will not be applied (for example, if a [Call REST Service](/refguide/call-rest-action/) returns an error, any data the REST call retrieved will not be kept). The error handling only applies to activities which happened before the activity which errors.

Reviewer: > In the examples we have an activity which generates the error. In a real app, however, an error can be generated by any activity.

I do not see the contrast between 'an activity' and 'any activity'. Maybe change 'an activity' to 'a Microflow call action' or 'a specific activity'?

Author: Good point. How about:
```suggestion
In the following subsections, we introduce the functionality of each error handling option using simple examples. In the examples we have a microflow call which generates the error. In a real app, however, an error can be generated by any activity. No results from the activity which fails will be available (for example, if a [Call REST Service](/refguide/call-rest-action/) returns an error, any data the REST call retrieved will not be kept). The error handling only applies to activities which happened before the activity which errors.
```

Reviewer: ~~an~~ a microflow call

> In the following subsections, we introduce the functionality of each error handling option using simple examples. In the examples we have an activity which generates the error. In a real app, however, an error can be generated by any activity. Any changes which happen within the activity which fails will not be applied (for example, if a [Call REST Service](/refguide/call-rest-action/) returns an error, any data the REST call retrieved will not be kept). The error handling only applies to activities which happened before the activity which errors.

Reviewer: > The error handling only applies to activities which happened before the activity which errors.

Isn't this redundant? I can't see what the alternative would be.

Author: I added this because someone pointed out that people could think that if something like a change object with commit fails on the commit part, people might think that the change object bit of the activity still happens. This is supposed to say that nothing that happens within the activity will still be kept.

Reviewer: > The error handling only applies to activities which happened before the activity which errors.

I interpreted this to mean that error handling only applies to activities that come in the flow before the activity with error handling. I do not see how to read it so it applies to activities with 'multiple steps', like the change activity. Can we rephrase it?

In this example, the error handling in **Microflow 2** is the default: **Rollback**.

{{< figure src="/attachments/refguide/modeling/application-logic/microflows-and-nanoflows/microflows/error-handling-in-microflows/default-roll-back.png" width="600px" class="no-border" alt="Two Mendix microflows. The first microflow, labeled 'Microflow 1,' starts with a green circle labeled 'savepoint 1,' proceeds to 'Create Customer' (NewCustomer, Customer), then calls 'Microflow 2,' and ends with a red circle. The second microflow, labeled 'Microflow 2,' starts with a green circle, proceeds to 'Create Order' (NewOrder, Order), then to 'GenerateError' (Variable, Integer/Long), and ends with a red circle." >}}

When **Microflow 1** starts with a button click, a savepoint is created at the very beginning of **Microflow 1**. When an error occurs in **Microflow 2**, the default error handling is to roll back all changes to the state of this savepoint (so changes made in **Create Order** and in **Create Customer** are both reverted.), **Microflow 2** terminates, then **Microflow 1** terminates, the error is logged, and a system error message is shown to the end-user.

{{% alert color="warning" %}}

> After a rollback, previously committed objects remain marked as unchanged in memory, even though their database state has been reverted. See [How Commits Work](/refguide/committing-objects/#how-commits-work) in *Commit Object(s)* to see the implications of this.

Reviewer: > After a rollback

Explicit '_database_ rollback'?

Author: I don't think we should say that here. Mendix uses the term "Rollback" not "Database Rollback" and I think we should be consistent with that. Unless you want to change the UX and say "Custom with Database Rollback" for the error handling option?

When **Microflow 1** starts with a button click, a savepoint is created at the very beginning of **Microflow 1**. When an error occurs in **Microflow 2**, everything rolls back to the state of this savepoint because the error handling is set to **Custom with rollback**. Hence, changes made in **Create Order** and in **Create Customer** are both reverted. A custom error is logged using a **Log message** activity. After that, the error event throws a new error (with the same information as the original error) to terminate **Microflow 2**, and **Microflow 1** will terminate because there is no custom error handing there.

### Error Handling - Custom without Rollback

> Setting **Custom without rollback** will not stop data changes within the activity which failed from being rolled back. For example, if a [Call REST Service](/refguide/call-rest-action/) returns an error, any data the REST call retrieved will not be kept.

Reviewer: > will not stop data changes within the activity which failed from being rolled back.

I find this sentence difficult to understand. Maybe interpreting this as 'rolling back' assumes too much. E.g. if the error handling is configured on a **Call microflow** action, then some of its side effects might be undone, some might not. In the given example of **Call REST Service** nothing is 'rolled back' per se, it's just that the return value is discarded.

Author: This is the same thing as I tried to say before. How about:

```suggestion
Setting **Custom without rollback** will not stop data changes within the actual activity which failed being lost. For example, if a [Call REST Service](/refguide/call-rest-action/) returns an error, any data the REST call retrieved will not be kept.
```

When **Microflow 1** starts with a button click, a savepoint is created at the very beginning of **Microflow 1**. Another savepoint is created right before **GenerateError** because the error handling is set to **Custom without rollback**. When an error occurs in **Microflow 2**, changes made in **Create Order** are at this moment still kept because of the savepoint right before **GenerateError**. A custom error is logged using a **Log message** activity. After that, the error event throws a new error (with the same information as the original error) to terminate **Microflow 2** and, because this error is not handled in **Microflow 1**, rolls back everything to the state of **savepoint 1** which is at the very beginning of **Microflow 1**. Hence, changes made in **Create Customer** and in **Create Order** are both reverted.

{{% alert color="info" %}}

> Using custom error handling without rollback and ending the error handling flow with an error event here makes it possible to create a custom log message with details about the **NewOrder** (if you used error handling with rollback, the information about the **NewOrder** would not be available when creating the custom log message). However, after logging this message, everything will still be rolled back as the error event is not handled in **Microflow 1**.  

Reviewer: > (if you used error handling with rollback, the information about the **NewOrder** would not be available when creating the custom log message)

I do not understand why this would be the case. The database state is rolled back to the savepoint, but the data in the **NewOrder** object should still be available.

I have asked it within my team, waiting for response.

Reviewer: We have no idea why this data would not be available. I tried within a test project, and I can read the attributes from the object. Or is it meant that retrieving the data from the database won't show the changed values? Which would be logical, as they were rolled back 😃

Author: You need to log the information while you still have the object in context. So you can't error and then log it in the calling microflow. That is what this is saying: you can do the custom error handling to log information from the object, and then rethrow the error. But it sounds like there is a confusion about what is being rolled back here. Perhaps we can make it clearer:

```suggestion
Using custom error handling and ending the error handling flow with an error event here makes it possible to create a custom log message with details about the **NewOrder**. However, after logging this message, everything will still be rolled back as the error event is not handled in **Microflow 1**.  
```

* Always use a log activity to print the error message and stack trace.
* Add custom error handling for activities that depend on third-party systems, as these activities might fail and usually provide detailed information on the cause of failure. With custom error handling, you should log the object that might have caused the error, the error type and message, or any responses from REST, SOAP, or Java action calls.
* Where possible, avoid using the **Continue** option and use **Custom without rollback** instead.
* Do not try to commit objects which were rolled back, Mendix will no longer see them as being changed. See [How Commits Work](/refguide/committing-objects/#how-commits-work) in *Commit Object(s)* for more information about this.
* Do not overdo it – you can specify a lot of complicated error handling combinations, but this makes it more difficult (and slower) for the Mendix Runtime to evaluate the microflow, and it also makes it more difficult to predict the exact behavior in case of an exception.

## What Error Handling Gives You

### Consistent Data

> Everything that happens in your app happens in a transaction. By default either all instructions are completed or nothing is completed. If you don't specify any error handling and the microflow you are trying to run encounters an error, it will appear that nothing has changed. That means that all the objects you created or changed will be reverted, and the end-user will get an error. This ensures that processes and data remain consistent. 

Reviewer 3: We don't tend to use "instructions", maybe "activities"? But this only applies to local objects, any calls to external systems cannot be reverted by a transaction rollback.

Author: Added this information in 89a5f3336953ea3f5b8ba388e6b806e496d294c4

### Isolated Changes

While updating or creating your objects, you do not want other end-users to see information until your microflow has finished executing. 

To ensure that every end-user or process can only see persisted data, data changed in a microflow is only available within that microflow. None of the changes made inside the microflow will be available outside the microflow, not even to the end-user that initiated the microflow. The information will only be available to all end-users of the application once the microflow has successfully completed all the activities.

### Protection from Parallel Updates

> When an object is updated, your app will place a lock on that object until the transaction/microflow ends. While the microflow is running, nothing else can read or write that same object and anything that attempts to do so will have to wait. This lock is released automatically as soon as the microflow ends, allowing any waiting processes to continue normally.

Reviewer 3: You can read an object that was updated and therefore locked by another transaction. Locks do not block reads, only other writes. The value you will see will not include any changes by uncommitted transactions.

Author: Tried to clarify in 89a5f3336953ea3f5b8ba388e6b806e496d294c4

* Do not overdo it – you can specify a lot of complicated error handling combinations, but this makes it more difficult (and slower) for the Mendix Runtime to evaluate the microflow, and it also makes it more difficult to predict the exact behavior in case of an exception.

## What Error Handling Gives You

### Consistent Data

Everything that happens within your app happens in a transaction. By default either all activities are completed or nothing is completed. If you don't specify any error handling and the microflow you are trying to run encounters an error, it will appear that nothing has changed. That means that all the objects you created or changed will be reverted, and the end-user will get an error. This ensures that processes and data remain consistent.

{{% alert color="info" %}}
This applies to data held within your app and its database. Calls to external systems will not be reversed if a transaction errors. You have to handle these cases specifically.
{{% /alert %}}

### Isolated Changes

While updating or creating your objects, you do not want other end-users to see updated information until your microflow has finished executing. 

> To ensure that every end-user or process can only see persisted data, data changed in a microflow is only available within that microflow. None of the changes made inside the microflow will be available outside the microflow, not even to the end-user that initiated the microflow. The information will only be available to all end-users of the application once the microflow has successfully completed all the activities.

Author: To address https://github.com/mendix/docs/pull/10899/changes#r3063102012 and try to explain how information is isolated:

```suggestion
To ensure that every end-user or process can only see persisted data, data changed and committed in a microflow is not made available to all end-users of the application until the top-level microflow has successfully completed all the activities. Before then, only the end-user running the microflow has access to the data. 
```

Reviewer: I think this is fine.

When an object is updated, your app will place a write lock on that object until the transaction/microflow ends. While the microflow is running, nothing else can write to that same object and anything that attempts to do so will have to wait. This lock is released automatically as soon as the microflow ends, allowing any waiting processes to continue normally.

{{% alert color="info" %}}

> This doesn't prevent two end-users editing the same object. Two end-users can read the same object and change it. However, neither will see the changes that the other has made. They can both commit the changes and the change committed second will overwrite the changes committed first.

Reviewer: > They can both commit the changes and the change committed second will overwrite the changes committed first.

Or it might error out, depending on whether [optimistic locking](https://docs.mendix.com/refguide/optimistic-locking/) is enabled.

Author: Good point - this was written before optimistic locking was introduced. Perhaps we should say:

```suggestion
This doesn't prevent two end-users editing the same object. Two end-users can read the same object and change it. However, neither will see the changes that the other has made. They can both commit the changes and the result depends on whether [optimistic locking](/refguide/optimistic-locking/) in implemented.

* Without optimistic locking, the change committed second will overwrite the changes committed first.
* WIth optimistic locking, the change committed second will throw an error.
```

Reviewer: `[optimisti locking]` -> `[optimistic locking]`

I would say 'enabled' instead of 'implemented'.
